rust/src/index/prefilter.rs

```diff
@@ -32,11 +32,37 @@ use crate::Dataset;
 /// is just deleted rows.
 pub struct PreFilter {
     dataset: Arc<Dataset>,
+    has_deletion_vectors: bool,
+    has_missing_fragments: bool,
 }
 
 impl PreFilter {
     pub fn new(dataset: Arc<Dataset>) -> Self {
-        Self { dataset }
+        let dataset_ref = dataset.as_ref();
+        let mut has_fragment = Vec::new();
+        let mut has_deletion_vectors = false;
+        has_fragment.resize(
+            (dataset
+                .manifest
+                .max_fragment_id()
+                .map(|id| id + 1)
+                .unwrap_or(0)) as usize,
+            false,
+        );
+        for frag in dataset_ref.manifest.fragments.iter() {
+            has_fragment[frag.id as usize] = true;
+            has_deletion_vectors |= frag.deletion_file.is_some();
+        }
+        let has_missing_fragments = has_fragment.iter().any(|&x| !x);
```

Comment:
What does missing fragment mean here? It is not clear to me why we track missing fragments ( i assume the holes in fragment ids?) for the deletion filter.

Reply:
IIRC, `has_missing_fragments` means that "there is no deleted rows ever happen to this dataset".  while we only need to skip filtering if there is no active deletion vector?

Reply:
> What does missing fragment mean here? It is not clear to me why we track missing fragments ( i assume the holes in fragment ids?) for the deletion filter.

If you delete all the rows in a fragment, we just remove the entire fragment. But it might still be present in the index, so we need to filter the row ids out when we get the top `k` matches.

Reply:
This will be problematic even we re-indexing a dataset, isn't it?

Reply:
Yeah we might want to add a list of fragments contained in the index, then we can use that list instead.